src/NServiceBus.Hosting.Windows/GenericHost.cs

```diff
@@ -70,27 +71,67 @@ public void Start()
         }
 
         /// <summary>
-        ///     Finalize
+        /// Shut down bus or give up after 30 seconds, so host won't hang indefinitely.
         /// </summary>
         public void Stop()
         {
-            wcfManager.Shutdown();
+            Task disposeTask = DisposeBusTask();
 
-            if (bus != null)
-            {
-                bus.Dispose();
+            StartShutdownReporter();
+            var timeoutTask = Task.Delay(TimeSpan.FromSeconds(30));
+
+            var finishedResult = Task.WhenAny(disposeTask, timeoutTask).GetAwaiter().GetResult();
 
-                bus = null;
+            if (finishedResult.Equals(timeoutTask))
+            {
+                LogManager.GetLogger<GenericHost>().Error("The host failed to stop with-in the time allowed (30s).");
```

Comment:
I heard you're not allowed to use readonly :)
You're right, but it was already like this. Anyone any idea if there's a reason why it's not using ILog ?